DOCS-12248 Google Pub/Sub 1.0 (#1648)
mulesoft/docs-connectors: google-pubsub/1.0/modules/ROOT/pages/google-pubsub-connector-reference.adoc

==== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
|Name | String | Name for this configuration. Connectors reference the configuration with this name. | | x
| Connection a| <<Config_Connection, PubSub Connection>>
 | Connection types for this configuration. | | x
| Default Project Id a| String |  Globally unique identifier for your project. Used in operations. It can be overridden at the config level. |  |
| Default Subscription Name a| String |  Default subscription for your project. Used in operations. It can be overridden at the config level. |  |
| Default Topic Name a| String |  Default topic for your project. Used in operations. It can be overridden at the config level. |  |
| Name a| String |  ID used to reference this configuration. |  | x
| Expiration Policy a| <<ExpirationPolicy>> |  Configures the minimum amount of time that a dynamic configuration instance can remain idle before Mule considers it eligible for expiration. |  |
|===

==== Connection Types
[[Config_Connection]]
===== PubSub Connection

====== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required

> | Private key ID a| String |  ID of the private key for the service account. It is a part of the service-credentials created in PubSub. |  | x

Reviewer: The ID is part of...
ditto for the rest of the rows

* <<RepeatableFileStoreIterable>>
* non-repeatable-iterable |  Configures how Mule processes streams. The default is to use repeatable streams. |  |
| Target Variable a| String |  Name of the variable that stores the operation's output. |  |
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] |
| Reconnection Strategy a| * <<Reconnect>>
* <<ReconnectForever>> |  Retry strategy in case of connectivity errors. |  |
|===

=== Output

[%autowidth.spread]
|===
|Type |Array of <<Snapshot>>
|===

=== For Configurations

* <<Config>>

=== Throws

* PUBSUB:BAD_GATEWAY
* PUBSUB:CANCELLED
* PUBSUB:DEADLINE_EXCEEDED
* PUBSUB:FAILED_PRECONDITION
* PUBSUB:INVALID_ARGUMENT
* PUBSUB:NOT_FOUND
* PUBSUB:PERMISSION_DENIED
* PUBSUB:UNAVAILABLE

[[PatchSnapshot]]
== Patch Snapshot
`<pubsub:patch-snapshot>`

Updates an existing snapshot.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Snapshot a| Any |  Snapshot object with updated fields. |  #[payload] |

> | Update Mask a| String |  Indicates which fields in the provided snapshot to update. Must be specified and not empty. This is a comma-separated list of fully qualified names of fields. |  | x

Reviewer: You can remove the second sentence since the field is required.

This is a comma-separated list of fully qualified names of fields. > This is a comma-separated list of fully qualified field names. (I think this works)

Author: Removed the second sentence. Kept "fully qualified names" since that is the actual term name

|Type |<<Snapshot>>
|===

=== For Configurations

* <<Config>>

=== Throws

* PUBSUB:BAD_GATEWAY
* PUBSUB:CANCELLED
* PUBSUB:CONNECTIVITY
* PUBSUB:DEADLINE_EXCEEDED
* PUBSUB:FAILED_PRECONDITION
* PUBSUB:INVALID_ARGUMENT
* PUBSUB:NOT_FOUND
* PUBSUB:PERMISSION_DENIED
* PUBSUB:RETRY_EXHAUSTED
* PUBSUB:UNAVAILABLE

[[PublishMessage]]
== Publish Message
`<pubsub:publish-message>`

Publishes a single message.

This operation is non-blocking. When batching is enabled, the message is
stored only locally until any of the batching restrictions are reached.

This operation may lead to high throughput. When using the *Ordering Key*
field, processing messages in the order they were called is not guaranteed,
especially in threaded environments. If you must maintain a strict order,
set *Max Concurrency* to `1` in your source flow settings and add a delay in
the for-each loops that contain order critical *Publish* calls.

This operation returns the `messageId`, which is the unique identifier of the
sent message in the topic.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x

> | Ordering Key a| String |  If not empty, the ordering key identifies related messages in which the publish order must be respected. This key is respected when the subscriber enables message ordering. |  |

Reviewer: ```suggestion
| Ordering Key a| String |  If specified, this field identifies related messages in which the publish order must be respected when the subscriber enables message ordering. |  |
```

| Message a| Binary |  Message content published to the topic. If not empty, the message must contain at least one attribute. |  |
| Attributes a| Object |  Attributes for this message. If empty, the message must contain non-empty data. Can be used to filter messages on the subscription. |  |
| Target Variable a| String |  Name of the variable that stores the operation's output. |  |
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] |
| Reconnection Strategy a| * <<Reconnect>>
* <<ReconnectForever>> |  Retry strategy in case of connectivity errors. |  |
|===

=== Output

[%autowidth.spread]
|===
| Type | String
|===

=== For Configurations

* <<Config>>

=== Throws

* PUBSUB:BAD_GATEWAY
* PUBSUB:CANCELLED
* PUBSUB:CONNECTIVITY
* PUBSUB:DEADLINE_EXCEEDED
* PUBSUB:FAILED_PRECONDITION
* PUBSUB:INVALID_ARGUMENT
* PUBSUB:MAX_OUTSTANDING_BATCH_SIZE_REACHED
* PUBSUB:MAX_OUTSTANDING_ELEMENT_COUNT_REACHED
* PUBSUB:NOT_FOUND
* PUBSUB:PERMISSION_DENIED
* PUBSUB:RETRY_EXHAUSTED
* PUBSUB:UNAVAILABLE

[[SeekMessages]]
== Seek Messages
`<pubsub:seek-messages>`

> Supports bulk acknowledging or un-acknowledging of messages to a given snapshot or point of time based on the seekTarget provided. Seeking to a time marks every message received by Pub/Sub before the time as acknowledged, and all messages received after the time as unacknowledged. You can replay and reprocess previously acknowledged messages when seeking to a time in the past or purge messages when seeking to a time in the future.

Reviewer: ```suggestion
Supports the bulk acknowledging or un-acknowledging of messages to a given snapshot or point of time based on the provided seek target. 

Seeking to a time marks every message received by Pub/Sub before the time as acknowledged, and all messages received after the time as unacknowledged. You can replay and reprocess previously acknowledged messages when seeking to a time in the past or purge messages when seeking to a time in the future.
```

Reviewer: A general comment with this material. It is complex, which makes it tough to word. It would be clearer if you separate the details that apply to seeking by snapshot and seeking to a point of time. The details should follow the order in which the features are introduced:

Supports the bulk acknowledging or un-acknowledging of messages to a point in time or to a snapshot

Seeking to a point in time...

Seeking to a snapshot enables you to...

If you seek to a timestamp using a subscription with a filter, the Pub/Sub service redelivers only the messages in the timestamp that match the filter of the subscription making the seek request. The snapshot feature enables you to capture the message acknowledgment state of a subscription. Once a snapshot is created, it retains:

* all messages that were unacknowledged in the source subscription at the time of the snapshot's creation.
* any messages published to the topic thereafter. You can replay these unacknowledged messages by using a snapshot to seek to any of the topic's subscriptions.

If you seek to a snapshot using a subscription with a filter, the Pub/Sub service redelivers only the messages in the snapshot that match the filter of the subscription making the seek request.

> 

Reviewer: Once a snapshot is created, it retains all messages that were unacknowledged in the source subscription at the time of the snapshot's creation and any messages published to the topic thereafter. You can replay these unacknowledged messages by using a snapshot to seek to any of the topic's subscriptions.

If you seek to a snapshot using a subscription with a filter, the Pub/Sub service redelivers only the messages in the snapshot that match the filter of the subscription making the seek request.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Config Ref a| ConfigurationProvider |  Name of the configuration used to execute this component. |  | x
| Project ID a| String |  Globally unique identifier for your project. |  |
| Subscription Name a| String |  Project-wide unique identifier for the subscription. |  |
| Snapshot Name a| String |  Snapshot to seek to. The snapshot's topic must be the same as that of the provided subscription. |  |
| Timestamp a| DateTime |  Specific time to seek to. Messages retained in the subscription that were published before this time are marked as acknowledged, and messages retained in the subscription that were published after this time are marked as unacknowledged.

> Note that this operation affects only messages retained in the subscription (configured by the combination of *messageRetentionDuration* and *retainAckedMessages*), for example, if the time corresponds to a point before the message retention window (or to a point before the system's notion of the subscription creation time), only retained messages are marked as unacknowledged, and already-expunged messages will not be restored. A timestamp is in RFC3339 UTC Zulu format, with nanosecond resolution and up to nine fractional digits, such as 2014-10-02T15:01:23Z and 2014-10-02T15:01:23.045123456Z. |  |

Reviewer: If possible, use  *Message Retention Duration* and *Retain Acked Messages* if these are field names in the connector config.

Reviewer: ```suggestion
This operation affects only messages retained in the subscription (configured by the combination of *messageRetentionDuration* and *retainAckedMessages*). For example, if the time corresponds to a point before the message retention window or to a point before the subscription creation time, only retained messages are marked as unacknowledged and already-expunged messages are not restored. A timestamp is in RFC3339 UTC Zulu format, with nanosecond resolution and up to nine fractional digits, such as 2014-10-02T15:01:23Z and 2014-10-02T15:01:23.045123456Z. |  |
```

| Reconnection Strategy a| * <<Reconnect>>
* <<ReconnectForever>> |  Retry strategy in case of connectivity errors. |  |
|===

=== For Configurations

* <<Config>>

=== Throws

* PUBSUB:BAD_GATEWAY
* PUBSUB:CANCELLED
* PUBSUB:CONNECTIVITY
* PUBSUB:DEADLINE_EXCEEDED
* PUBSUB:FAILED_PRECONDITION
* PUBSUB:INVALID_ARGUMENT
* PUBSUB:NOT_FOUND
* PUBSUB:PERMISSION_DENIED
* PUBSUB:RETRY_EXHAUSTED
* PUBSUB:UNAVAILABLE

== Sources

[[MessageListener]]
== On message listener
`<pubsub:message-listener>`

> Runnable message receiver, using Pub/Sub Subscriber.

Reviewer: Will users know what a runnable message receiver is? 
Message receiver for Pub/Sub subscriptions???

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Consumer count a| Number |  Provides a specified amount of executor service for processing messages. |  5 |
| Config Ref a| ConfigurationProvider |  Name of the configuration used to execute this component. |  | x
| Primary Node Only a| Boolean |  Determines whether to execute this source on only the primary node when running Mule instances in a cluster. |  |
| Streaming Strategy a| * <<RepeatableInMemoryStream>>
* <<RepeatableFileStoreStream>>
* non-repeatable-stream |  Configures how Mule processes streams. The default is to use repeatable streams. |  |
| Redelivery Policy a| <<RedeliveryPolicy>> |  Defines a policy for processing the redelivery of the same message. |  |
| Project ID a| String |  Globally unique identifier for your project. |  |
| Subscription Name a| String |  Project-wide unique identifier for the subscription. |  |
| Limit Exceeded Behavior a| Enumeration, one of:

** THROW_EXCEPTION
** BLOCK

> ** IGNORE |  If LimitExceededBehavior is not set to `IGNORE`, the flow control enables you to have control over the behavior of receiving messages by specifying the limits for *maxOutstandingRequestSize* and *maxOutstandingElementCount*. If the memory size or the amount of outstanding elements (messages where no Ack() or Nack() was performed) exceeds specified limits, the receiver won't receive more messages when LimitExceedBehavior is set to `BLOCK`, or will throw a FlowControlException when LimitExceedBehavior is set to `THROW_EXCEPTION`. |  BLOCK |

Reviewer: See if you can rewrite this using a bullet for each possible value. I'll look at it again afterward.

Enables you to control the behavior of receiving messages:

* THROW_EXCEPTION...
* BLOCK...
* IGNORE...

| Message Digest Algorithm a| String | Secure hashing algorithm to use if the *Use Secure Hash* field is `true`. If the payload of the message is a Java object, Mule ignores this value and returns the value that the payload’s `hashCode()` returned. |  |
| Message Identifier a| <<RedeliveryPolicyMessageIdentifier>> | Defines which strategy is used to identify the messages. |  |
| Object Store a| ObjectStore | Configures the object store that stores the redelivery counter for each message.  |  |
|===

[[RedeliveryPolicyMessageIdentifier]]
=== Redelivery Policy Message Identifier

Configures how to identify a redelivered message and how to find out when the message was redelivered.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Use Secure Hash a| Boolean | If `true`, Mule uses a secure hash algorithm to identify a redelivered message. |  |
| Id Expression a| String | One or more expressions that determine when a message was redelivered. You can set this property only if the *Use Secure Hash* field is `false`. |  |
|===

[[Snapshot]]
=== Snapshot

> Snapshot resource. Use snapshots in seek operations, which manage message acknowledgments in bulk.

Reviewer: Is a snapshot only used in seek operations and only used to help manage acknowledgements in bulk or does it have other uses?

This type enables you to set the acknowledgment state of messages in an existing subscription to the state captured by a snapshot.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Name a| String | Name of the snapshot. |  |
| Topic a| String | Name of the topic from which the snapshot retains messages. |  |

> | Expire Time a| DateTime | Time that the snapshot is guaranteed to exist up until. A newly-created snapshot expires no later than seven days from the time of its creation. Its exact lifetime is determined at creation by the existing backlog in the source subscription. The lifetime of the snapshot is seven days, which is the age of the oldest unacked message in the subscription. Consider a subscription with an oldest unacked message that is three days old. If a snapshot is created from this subscription, the snapshot, which always captures this three day old backlog as long as the snapshot exists, will expire in four days. The service will refuse to create a snapshot that will expire in less than one hour after creation. A timestamp is in RFC3339 UTC Zulu format, with nanosecond resolution and up to nine fractional digits, such as 2014-10-02T15:01:23Z and 2014-10-02T15:01:23.045123456Z. |  |

Reviewer: Tough to understand. I think it would be clearer if you separate it into paragraphs. The way I do this for table entries is as follows:

{sp} +
{sp} +

Reviewer: Also, see if you can remove some of the asides. For example, "the snapshot, which always captures this three day old backlog as long as the snapshot exists, will expire in four days" - you can probably change to "the snapshot will expire in four days."

* <<RepeatableFileStoreIterable>>
* non-repeatable-iterable |  Configures how Mule processes streams. The default is to use repeatable streams. |  |
| Target Variable a| String |  Name of the variable that stores the operation's output. |  |
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] |
| Reconnection Strategy a| * <<Reconnect>>
* <<ReconnectForever>> |  Retry strategy in case of connectivity errors. |  |
|===

=== Output

[%autowidth.spread]
|===
|Type |Array of <<Snapshot>>
|===

=== For Configurations

* <<Config>>

=== Throws

* PUBSUB:BAD_GATEWAY
* PUBSUB:CANCELLED
* PUBSUB:DEADLINE_EXCEEDED
* PUBSUB:FAILED_PRECONDITION
* PUBSUB:INVALID_ARGUMENT
* PUBSUB:NOT_FOUND
* PUBSUB:PERMISSION_DENIED
* PUBSUB:UNAVAILABLE

[[PatchSnapshot]]
== Patch Snapshot
`<pubsub:patch-snapshot>`

Updates an existing snapshot.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Snapshot a| Any |  Snapshot object with updated fields. |  #[payload] |

> | Update Mask a| String |  Indicates which fields in the provided snapshot to update. This is a comma-separated list of fully qualified names of fields. |  | x

Reviewer: ```suggestion
| Update Mask a| String |  Indicates which fields in the provided snapshot to update. This is a comma-separated list of fully qualified field names. |  | x
```

Author: Keeping "fully qualified names" since that is the name of the actual term

[[Snapshot]]
=== Snapshot

> Snapshot resource. You may use snapshots in seek operations, which manage message acknowledgments in bulk.

Reviewer: may > can

| Field | Type | Description | Default Value | Required
| Name a| String | Name of the snapshot. |  |
| Topic a| String | Name of the topic from which the snapshot retains messages. |  |
| Expire Time a| DateTime |

> Time that the snapshot is guaranteed to exist up until. A newly-created snapshot

Reviewer: The parenthetical expression confuses things. Can you write something like this? (I also reworded the first sentence a bit to make it less awkward.)

Time length for which the snapshot is guaranteed to exist. The lifetime for a newly-created snapshot is based on the oldest unacked message in the source subscription's backlog, however, the snapshot can exist for a maximum of seven days.

Consider...
